File: samples/snippets/fsharp/VS_Snippets_CLR_System/System.ArgumentOutOfRangeException/fs/Race2.fs
module Race2

// <Snippet12>
open System.Collections.Concurrent
open System.Threading

type Continent =
    { Name: string
      Population: int
      Area: decimal }

let continents = ConcurrentBag<Continent>();
let mutable msg = ""

let names = 
    [ "Africa"; "Antarctica"; "Asia"
      "Australia"; "Europe"; "North America"
      "South America" ]

let gate = new CountdownEvent(names.Length)

let populateContinents obj = 
    let name = string obj
    lock msg (fun () -> 
        msg <- msg + $"Adding '{name}' to the list.\n" )

    // Sleep to simulate retrieving remaining data.
    let continent = 
        { Name = name
          Population = 0
          Area = 0M }
    Thread.Sleep 25
    continents.Add continent
    gate.Signal() |> ignore

// Populate the list.
for name in names do
    let th = Thread(ParameterizedThreadStart populateContinents)
    th.Start name

// Display the list.
gate.Wait();
printfn $"{msg}\n"

let arr = continents.ToArray();
for i = 0 to names.Length - 1 do
    let continent = arr[i]
    printfn $"{continent.Name}: Area: {continent.Population}, Population {continent.Area}"

// The example displays output like the following:
//       Adding 'Africa' to the list.
//       Adding 'Antarctica' to the list.
//       Adding 'Asia' to the list.
//       Adding 'Australia' to the list.
//       Adding 'Europe' to the list.
//       Adding 'North America' to the list.
//       Adding 'South America' to the list.
//
//
//       Africa: Area: 0, Population 0
//       Antarctica: Area: 0, Population 0
//       Asia: Area: 0, Population 0
//       Australia: Area: 0, Population 0
//       Europe: Area: 0, Population 0
//       North America: Area: 0, Population 0
//       South America: Area: 0, Population 0
// </Snippet12>

File: snippets/csharp/System/ArgumentOutOfRangeException/Overview/Race2.cs
// <Snippet12>
using System;
using System.Collections.Concurrent;
using System.Threading;

public class ContinentD
{
    public string? Name { get; set; }
    public int Population { get; set; }
    public decimal Area { get; set; }
}

public class Example12
{
    static ConcurrentBag<ContinentD> ContinentDs = new();
    static CountdownEvent? gate;
    static string msg = string.Empty;

    public static void Main()
    {
        string[] names = { "Africa", "Antarctica", "Asia",
                         "Australia", "Europe", "North America",
                         "South America" };
        gate = new(names.Length);

        // Populate the list.
        foreach (string name in names)
        {
            var th = new Thread(PopulateContinentDs);
            th.Start(name);
        }

        // Display the list.
        gate.Wait();
        Console.WriteLine(msg);
        Console.WriteLine();

        var arr = ContinentDs.ToArray();
        for (int ctr = 0; ctr < names.Length; ctr++)
        {
            var ContinentD = arr[ctr];
            Console.WriteLine($"{ContinentD.Name}: Area: {ContinentD.Population}, Population {ContinentD.Area}");
        }
    }

    private static void PopulateContinentDs(object? obj)
    {
        string? name = obj?.ToString();
        lock (msg)
        {
            msg += $"Adding '{name}' to the list.\n";
        }
        var ContinentD = new ContinentD() { Name = name };
        // Sleep to simulate retrieving remaining data.
        Thread.Sleep(25);
        ContinentDs.Add(ContinentD);
        gate?.Signal();
    }
}
// The example displays output like the following:
//       Adding 'Africa' to the list.
//       Adding 'Antarctica' to the list.
//       Adding 'Asia' to the list.
//       Adding 'Australia' to the list.
//       Adding 'Europe' to the list.
//       Adding 'North America' to the list.
//       Adding 'South America' to the list.
//
//
//       Africa: Area: 0, Population 0
//       Antarctica: Area: 0, Population 0
//       Asia: Area: 0, Population 0
//       Australia: Area: 0, Population 0
//       Europe: Area: 0, Population 0
//       North America: Area: 0, Population 0
//       South America: Area: 0, Population 0
// </Snippet12>
